Code/GraphMol/CIPLabeler/CIPLabeler.cpp

```diff
@@ -0,0 +1,203 @@
+//
+//
+//  Copyright (C) 2020 Schrödinger, LLC
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <algorithm>
+
+#include <GraphMol/RDKitBase.h>
+
+#include "CIPLabeler.h"
+#include "CIPMol.h"
+#include "configs/Sp2Bond.h"
+#include "configs/Tetrahedral.h"
+
+#include "rules/Rules.hpp"
+#include "rules/Rule1a.h"
+#include "rules/Rule1b.h"
+#include "rules/Rule2.h"
+#include "rules/Rule3.h"
+#include "rules/Rule4a.h"
+#include "rules/Rule4b.h"
+#include "rules/Rule4c.h"
+#include "rules/Rule5New.h"
+#include "rules/Rule6.h"
+
+namespace RDKit {
+namespace CIPLabeler {
+
+namespace {
+std::vector<Configuration *> findConfigs(CIPMol &mol) {
+  std::vector<Configuration *> configs;
+
+  for (auto &atom : mol.atoms()) {
+    auto chiraltag = static_cast<int>(atom->getChiralTag());
+    if (chiraltag > 0) {
+      auto carriers = std::vector<Atom *>();
+      carriers.reserve(4);
+      for (auto &bond : mol.getBonds(atom)) {
+        auto nbr = bond->getOtherAtom(atom);
+        carriers.push_back(nbr);
+      }
+      if (carriers.size() < 4) {
+        // Implicit H -- use the central atom instead of a dummy H
+        carriers.push_back(atom);
+      }
+
+      // RDKit uses SMILES parity, and centres requires SDF parity, which
+      // is opposite.
+      auto parity = chiraltag ^ 0x3;
```

Comment:
No, this comes from the original labeler code. `parity ^ 0x3` is used somewhere else to flip the parity. I just used it here for the conversion.

I'm not 100% convinced that RDKit and SDF parities are always opposite, but it has worked well so far.

Reply:
I don't think this is generally true.
SDF parity is done using the numbering of the neighbors: you point the highest number neighbor away from you and then look to see which direction you have to rotate to move the lowest numbered neighbor the next lowest numbered neighbor, CW is parity 1, CCW is parity 2.
RDKit parity, for dumb historic reasons, uses the order of the bonds instead of the order of the atoms. You look *down* the highest priority (=first) bond and then look at the rotation direction to move from the atom at thend of the second bond to atom at the end of the third bond. The relative numbering of the bonds often tracks the relative numbering of the atoms, but not always.

I just played with an actual physical model and I think that as long as the ordering of the bonds corresponds to the ordering of the atoms, the RDKit parity and the SDF parity are the same.

Reply:
Hum. That is as interesting thing, because, then, there's something that is yielding the exact opposite results for just the Tetrahedral configurations in the Validation suite.

That's going to be an interesting one to debug...

Reply:
Ok, it was easier than I expected: I just had the translation from `CW`/`CCW` to `LEFT`/`RIGHT` wrong.